### PR TITLE
Docker: improve caching, exclude more from build context, allow external RPC connections

### PR DESCRIPTION
**Pull request by coriolinus**
**Why change the builder base image?**

`phusion/baseimage`-based builds consistently failed for me in the `apt update` stage; some kind of problem with unsigned packages. I judged that we don't need `baseimage` because for a build stage, we don't have the problems that that package solves.

**Why make these changes at all?**

By copying external data into the image as late as possible, we maximize usage of Docker's cache. I made a few attempts to further improve caching which unfortunately didn't work; see 
86dcca2 and 9a93955 for details. Still, when building locally on my macbook, this consistently shaves off 5-10 minutes of build time.

Adding additional patterns to `.dockerignore` reduces the build context size from ~70Mb to ~5Mb. This only saves a few seconds of build time, but does prevent some spurious cache misses.

Adding the flag enabling RPC access to the node enables access from the outside world. A node which refuses to interact with anything external to itself is of severely limited utility.

### CODE REVIEW

**Comment by coriolinus** [comment]
[clabot:check]

**Comment by coriolinus** [comment]
To reviewers: I don't know what to do about this CI failure:

```
wasm source files changed but not the spec/impl version and the runtime
 binary blob. If changes made do not alter logic, just bump 'impl_version'.
 If they do change logic, bump 'spec_version' and rebuild wasm.
 source file directories:
 - runtime
 versions file: runtime/kusama/src/lib.rs
```

Given that I haven't changed any Rust code, the WASM binaries should not have changed. Is this failure spurious?

File: docker/Dockerfile [hunk 1]
```diff
@@ -1,40 +1,44 @@
...
+    rm -rf /usr/bin /usr/sbin
 
 USER polkadot
 EXPOSE 30333 9933 9944
 VOLUME ["/data"]
 
-CMD ["/usr/local/bin/polkadot"]
+COPY --from=builder /polkadot/target/$PROFILE/polkadot /usr/local/bin
+
+CMD ["/usr/local/bin/polkadot", "--rpc-external"]
```
Comment (bkchr) [inline docker/Dockerfile]: Why is this hard coded?

You will need other parameters as well.

File: .dockerignore [hunk 2]
```diff
@@ -1,2 +1,9 @@
 doc
 **/target
+docker/
+**/*.md
+**/*.txt
+.git/
```
Comment (bkchr) [inline .dockerignore]: We are extracting the git commit in the build process.

File: docker/Dockerfile (hunk 1 above)
Comment (coriolinus) [reply docker/Dockerfile]: It's straightforward to override externally, but I had a hard time imagining a use case for which external RPC access wasn't desired.

File: .dockerignore (hunk 2 above)
Comment (coriolinus) [reply .dockerignore]: Good point. I'll have to look into a way to work around that; I think that saving ~65mb of context is worth updating the mechanism by which we get the commit.

File: docker/Dockerfile (hunk 1 above)
Comment (bkchr) [reply docker/Dockerfile]: OKay.

File: .dockerignore (hunk 2 above)
Comment (coriolinus) [reply .dockerignore]: Though it's worth noting that even excluding `.git`, dockerized images appear to build properly:

```sh
~/paritytech/cumulus$ docker run --rm polkadot:prgn-dockerfile-caching /usr/local/bin/polkadot --version
polkadot 0.7.17-x86_64-linux-gnu
```

Comment (bkchr) [reply .dockerignore]: Yeah I know, we support building without a git ref. I think we just print a warning at compile time.

**Pull request merged**